### PR TITLE
fix: zoom level resets on window change after navigation

### LINKED ISSUE #5265: Flickering Zoom levels
### Bug description

Zoom/resize problem when bouncing between chrome and podman desktop. You'll see in the video, I define a zoom level, I switch to chrome, I close podman desktop and sometime the zoom level is reset (either bigger or smaller).
I also have the same behavior happening when I resize the podman desktop window.
Happens with or without an external monitor.
Reproduced on Linux too

https://github.com/containers/podman-desktop/assets/1636769/eb2d7588-dfd3-4b2a-8d52-e6ad8939c22b

(see in the steps to reproduce the explanations about the video)

### Operating system

macOS 14.2.1
windows 10

### Installation Method

Installer from website/GitHub releases

### Version

1.6.0

### Steps to reproduce

- I start at normal zoom level (just the one when starting the app)
- I bounce between chrome and PD
- Then I close with the "closing app button" on mac
- _BUG1 (0:11 --> 0:24)_ Then I reopen it by going to the tray menu, dashboard --> **Here it automatically Zoomed In (probably 2 or 3 levels)**
- (0:24 --> 0:29) Then I'm zooming in with keyboard shortcut --> working fine
- (0:29 --> 0:40) Then I'm moving around in the app and it works fine
- _BUG2 (0:40 --> 0:42)_ I'm zooming out. **On the first press on the keyboard short cut, it is zooming out by 2 or 3 levels**
- (0:42 --> 1:12) I'm again bouncing between chrome and my environment. Things are going fine
- _BUG3 (1:12 --> 1:15)_ I'm now resizing my podman desktop window. And you can see that **the zoom level change while I'm resizing, but I'm not changing anything manually.**

### Relevant log output

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by dgolovin**
### What does this PR do?

Switching routing mode to memory fixes the issue and according the doc here https://github.com/AlexxNB/tinro?tab=readme-ov-file#navigation-method

`By default, navigation uses the History API which allows you to have clean page URLs, although it needs some setup on the server side. Instead, you may choose to use hash or memory navigation methods.

There is no need to change links or paths in your app, everything else will still work the same.`

### Screenshot / video of UI

Before

https://github.com/user-attachments/assets/d8330e76-1c06-475f-80ed-cd0ba8b6a0ff

after

https://github.com/user-attachments/assets/3d202ce3-cab6-41b6-b4f2-84c0cfc3d2d2

### What issues does this PR fix or reference?

Fix #5265.

### How to test this PR?

Run in terminal to pull all relevant branches:

```
git remote add dg https://github.com/dgolovin
git remote fetch dg
git checkout i5265-switch-to-memory-mode
git pull dg i5265_navigation_fix --rebase
git pull dg allow-to-run-in-electron --rebase
```

Install dependencies, build and run podman-desktop in electron

```
yarn
yarn build
node_modules/.bin/electron .
```

When podman desktop window shows up zoom up window content and alternate page navigation and window resize. There should be no zoom changes.

- [ ] Tests are covering the bug fix or the new feature

### CODE REVIEW

**Review by deboer-tim — COMMENTED** [review]
It seems like there are at least three changes here:
- Changing from hash to memory mode.
- 'Moving' the dashboard from `/` to `/dashboard`. (why?)
- Using the navigation API in a few more places.

It's not clear to me that the second or third are related or necessary to fix this issue. If not, maybe these could be split into another PR(s)?

**Comment by dgolovin** [comment]
@deboer-tim Simple change that fixes the issues is changing mode to 'memory'. Which developers of tinro says should not affect anything, but it does. Hence all other changes to deal with issues after mode change to 'memory':
1. Dashboard would not show up at startup for some reason
2. windows.location.href used stops working

> 'Moving' the dashboard from / to /dashboard. (why?)

Probably is not needed, It just left after my experiments to fix (1)

Otherwise it seems to work everywhere else and it fixes zoom jumps on macOS and windows. Have to check if Linux platform could be affected too.

File: packages/renderer/src/App.svelte [hunk 1]
```diff
@@ -68,11 +68,12 @@ import WelcomePage from './lib/welcome/WelcomePage.svelte';
 import PreferencesNavigation from './PreferencesNavigation.svelte';
 import Route from './Route.svelte';
 
-router.mode.hash();
+router.mode.memory();
 
 //remember from where we come to preference pages
 let nonSettingsPage = '/';
 router.subscribe(function (navigation) {
+  console.log('->', navigation.url);
```
Comment (benoitf) [inline packages/renderer/src/App.svelte]:
```suggestion
```

File: packages/renderer/src/AppNavigation.svelte [hunk 2]
```diff
@@ -27,7 +30,9 @@ function clickSettings(b: boolean) {
   if (b) {
     exitSettingsCallback();
   } else {
-    window.location.href = '#/preferences/resources';
+    handleNavigation({
+      page: NavigationPage.RESOURCES,
+    });
   }
```
Comment (benoitf) [inline packages/renderer/src/AppNavigation.svelte]: maybe move this to a separate PR as it's unrelated

File: packages/renderer/src/lib/container/ContainerList.svelte [hunk 3]
```diff
@@ -59,7 +61,9 @@ $: updateContainers(containersInfo, globalContext, viewContributions, searchTerm
 
 function fromExistingImage(): void {
   openChoiceModal = false;
-  window.location.href = '#/images';
+  handleNavigation({
+    page: NavigationPage.IMAGES,
+  });
```
Comment (benoitf) [inline packages/renderer/src/lib/container/ContainerList.svelte]: separate PR

File: packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte [hunk 4]
```diff
@@ -250,7 +252,9 @@ function cleanupBuild(): void {
   }
 
   // redirect to the imlage list
-  window.location.href = '#/images';
+  handleNavigation({
+    page: NavigationPage.IMAGES,
+  });
```
Comment (benoitf) [inline packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte]: separate PR

File: packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte [hunk 5]
```diff
@@ -250,7 +252,9 @@ function cleanupBuild(): void {
   }
 
   // redirect to the imlage list
```
Comment (benoitf) [inline packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte]:
```suggestion
  // redirect to the image list
```

**Review by benoitf — COMMENTED** [review]
I think we could split this PR in 3

1. one easy/noop PR which just call handleNavigation rather than using window.location.href = 
2. one that introduce /dashboad
3. one that switch to memory usage

it's unclear the change around why $lastPage.path is renamed to /dashboard, it seems not to be the intended operation

File: packages/renderer/src/lib/container/ContainerList.svelte (hunk 3 above)
Comment (dgolovin) [inline packages/renderer/src/lib/container/ContainerList.svelte]:
```suggestion
  window.location.href = '#/images';
```

File: packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte (hunk 4 above)
Comment (dgolovin) [inline packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte]:
```suggestion
  window.location.href = '#/images';
```

File: packages/renderer/src/AppNavigation.svelte [hunk 6]
```diff
@@ -27,7 +30,9 @@ function clickSettings(b: boolean) {
   if (b) {
     exitSettingsCallback();
   } else {
-    window.location.href = '#/preferences/resources';
+    handleNavigation({
+      page: NavigationPage.RESOURCES,
+    });
```
Comment (dgolovin) [inline packages/renderer/src/AppNavigation.svelte]:
```suggestion
    window.location.href = '#/preferences/resources';
```

File: packages/renderer/src/App.spec.ts [hunk 7]
```diff
@@ -56,7 +56,7 @@ vi.mock('./lib/appearance/Appearance.svelte', () => ({
 
 beforeEach(() => {
   vi.resetAllMocks();
-  router.goto('/');
+  router.goto('/dashboard');
```
Comment (dgolovin) [inline packages/renderer/src/App.spec.ts]:
```suggestion
  router.goto('/');
```

File: packages/renderer/src/App.svelte [hunk 8]
```diff
@@ -110,7 +110,7 @@ window.events?.receive('navigate', (navigationRequest: unknown) => {
         <TaskManager />
         <SendFeedback />
         <ToastHandler />
-        <Route path="/" breadcrumb="Dashboard Page">
+        <Route path="/dashboard" breadcrumb="Dashboard Page">
```
Comment (dgolovin) [inline packages/renderer/src/App.svelte]:
```suggestion
        <Route path="/" breadcrumb="Dashboard Page">
```

File: packages/renderer/src/AppNavigation.svelte [hunk 9]
```diff
@@ -4,6 +4,8 @@
 import { onMount } from 'svelte';
 import type { TinroRouteMeta } from 'tinro';
 
+import { NavigationPage } from '/@api/navigation-page';
```
Comment (dgolovin) [inline packages/renderer/src/AppNavigation.svelte]:
```suggestion
```

File: packages/renderer/src/AppNavigation.svelte [hunk 10]
```diff
@@ -4,6 +4,8 @@
...
+
```
Comment (dgolovin) [inline packages/renderer/src/AppNavigation.svelte]:
```suggestion
```

File: packages/renderer/src/AppNavigation.svelte [hunk 11]
```diff
@@ -12,6 +14,7 @@ import SettingsIcon from './lib/images/SettingsIcon.svelte';
 import NavItem from './lib/ui/NavItem.svelte';
 import NavRegistryEntry from './lib/ui/NavRegistryEntry.svelte';
 import NavSection from './lib/ui/NavSection.svelte';
+import { handleNavigation } from './navigation';
```
Comment (dgolovin) [inline packages/renderer/src/AppNavigation.svelte]:
```suggestion
```

File: packages/renderer/src/AppNavigation.svelte [hunk 12]
```diff
@@ -36,7 +39,7 @@ function clickSettings(b: boolean) {
 <nav
   class="group w-leftnavbar min-w-leftnavbar flex flex-col hover:overflow-y-none bg-[var(--pd-global-nav-bg)] border-[var(--pd-global-nav-bg-border)] border-r-[1px]"
   aria-label="AppNavigation">
-  <NavItem href="/" tooltip="Dashboard" bind:meta={meta}>
+  <NavItem href="/dashboard" tooltip="Dashboard" bind:meta={meta}>
```
Comment (dgolovin) [inline packages/renderer/src/AppNavigation.svelte]:
```suggestion
```

File: packages/renderer/src/lib/container/ContainerList.svelte [hunk 13]
```diff
@@ -15,7 +15,9 @@ import { onDestroy, onMount } from 'svelte';
 import { get, type Unsubscriber } from 'svelte/store';
 import { router } from 'tinro';
 
+import { handleNavigation } from '/@/navigation';
```
Comment (dgolovin) [inline packages/renderer/src/lib/container/ContainerList.svelte]:
```suggestion
```

File: packages/renderer/src/lib/container/ContainerList.svelte [hunk 14]
```diff
@@ -15,7 +15,9 @@ import { onDestroy, onMount } from 'svelte';
...
 import type { ContainerInfo } from '/@api/container-info';
+import { NavigationPage } from '/@api/navigation-page';
```
Comment (dgolovin) [inline packages/renderer/src/lib/container/ContainerList.svelte]:
```suggestion
```

File: packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte [hunk 15]
```diff
@@ -9,7 +9,9 @@ import { get } from 'svelte/store';
 import type { Terminal } from 'xterm';
 
 import FileInput from '/@/lib/ui/FileInput.svelte';
+import { handleNavigation } from '/@/navigation';
```
Comment (dgolovin) [inline packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte]:
```suggestion
```

File: packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte [hunk 16]
```diff
@@ -9,7 +9,9 @@ import { get } from 'svelte/store';
...
 import { type BuildImageInfo, buildImagesInfo } from '/@/stores/build-images';
+import { NavigationPage } from '/@api/navigation-page';
```
Comment (dgolovin) [inline packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte]:
```suggestion
```

File: packages/renderer/src/lib/onboarding/Onboarding.svelte [hunk 17]
```diff
@@ -177,7 +176,7 @@ async function setActiveStep() {
   }
   telemetrySession.send(onboardings.map(o => o.extension).join(','), false);
   // if it reaches this point it means that the onboarding is fully completed and the user is redirected to the dashboard
-  router.goto($lastPage.path);
+  router.goto('/dashboard');
```
Comment (dgolovin) [inline packages/renderer/src/lib/onboarding/Onboarding.svelte]:
```suggestion
 router.goto($lastPage.path);
```

File: packages/renderer/src/lib/onboarding/Onboarding.svelte [hunk 18]
```diff
@@ -265,14 +264,15 @@ async function updateOnboardingStep() {
...
 }
 
 async function cancelSetup() {
   // TODO: it cancels all running commands
   // it redirect the user to the dashboard
   await cleanSetup(onboardings, globalContext);
   telemetrySession.send(onboardings.map(o => o.extension).join(','), true);
-  router.goto($lastPage.path);
+  router.goto('/dashboard');
```
Comment (dgolovin) [inline packages/renderer/src/lib/onboarding/Onboarding.svelte]:
```suggestion
  router.goto($lastPage.path);
```

File: packages/renderer/src/lib/welcome/WelcomePage.svelte [hunk 19]
```diff
@@ -37,6 +37,7 @@ onMount(async () => {
     welcomeUtils.updateVersion('initial');
     showWelcome = true;
   }
+  router.goto('/dashboard');
```
Comment (dgolovin) [inline packages/renderer/src/lib/welcome/WelcomePage.svelte]:
```suggestion
  router.goto('/');
```

**Review by dgolovin — COMMENTED** [review]
No need to rename route to dashboard, there is only one change required to work in memory mode. There should be routrer.goto('/') call in welcome page (see last change) or there would be empty screen after start.

File: packages/renderer/src/lib/onboarding/Onboarding.svelte [hunk 20]
```diff
@@ -177,7 +176,7 @@ async function setActiveStep() {
   }
   telemetrySession.send(onboardings.map(o => o.extension).join(','), false);
   // if it reaches this point it means that the onboarding is fully completed and the user is redirected to the dashboard
-  router.goto($lastPage.path);
+ router.goto($lastPage.path);
```
Comment (dgolovin) [inline packages/renderer/src/lib/onboarding/Onboarding.svelte]:
```suggestion
  router.goto($lastPage.path);
```

File: packages/renderer/src/lib/onboarding/Onboarding.svelte [hunk 21]
```diff
@@ -265,6 +264,7 @@ async function updateOnboardingStep() {
 
 function setDisplayCancelSetup(display: boolean) {
   displayCancelSetup = display;
+  router.goto('/dashboard');
```
Comment (dgolovin) [inline packages/renderer/src/lib/onboarding/Onboarding.svelte]:
```suggestion
  router.goto('/');
```

**Comment by dgolovin** [comment]
The route.goto('/') works fine, so there is no need to use '/dashboard'.

File: packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte [hunk 22]
```diff
@@ -249,7 +249,7 @@ function cleanupBuild(): void {
     buildImageInfo = undefined;
   }
 
-  // redirect to the imlage list
+  // redirect to the image list
```
Comment (dgolovin) [inline packages/renderer/src/lib/image/BuildImageFromContainerfile.svelte]:
```suggestion
  // redirect to the imlage list
```

**Comment by dgolovin** [comment]
smoke e2e failures should be fixed with https://github.com/containers/podman-desktop/pull/8542 merged.

**Comment by benoitf** [comment]
clicked on the rebase button

File: packages/renderer/src/lib/onboarding/Onboarding.svelte [hunk 23]
```diff
@@ -265,6 +265,7 @@ async function updateOnboardingStep() {
 
 function setDisplayCancelSetup(display: boolean) {
   displayCancelSetup = display;
+  router.goto('/');
```
Comment (lstocchi) [inline packages/renderer/src/lib/onboarding/Onboarding.svelte]: Why this change?  This function updates the visibility of the dialog where the user will be asked to: 1) keep going on the onboarding 2) stop the onboarding and leave page. If the user click on the second there is already a `router.goto`

Comment (dgolovin) [reply packages/renderer/src/lib/onboarding/Onboarding.svelte]: That is correct observation. This goto is not required.
```suggestion
```

**Review by benoitf — APPROVED** [review]
I thought I approved

**Pull request merged**